jetty-websocket/websocket-core-server/src/main/java/org/eclipse/jetty/websocket/core/server/WebSocketServerComponents.java

```diff
@@ -90,17 +90,20 @@ public static WebSocketComponents ensureWebSocketComponents(Server server, Servl
         if (server.contains(bufferPool))
             serverComponents.unmanage(bufferPool);
 
-        servletContext.setAttribute(WEBSOCKET_COMPONENTS_ATTRIBUTE, serverComponents);
-        LifeCycle.start(serverComponents);
-        servletContext.addListener(new ServletContextListener()
+        // Stop the WebSocketComponents when the ContextHandler stops.
+        ContextHandler contextHandler = Objects.requireNonNull(ContextHandler.getContextHandler(servletContext));
+        contextHandler.addManaged(serverComponents);
+        contextHandler.addEventListener(new LifeCycle.Listener()
         {
             @Override
-            public void contextDestroyed(ServletContextEvent sce)
+            public void lifeCycleStopping(LifeCycle event)
             {
-                LifeCycle.stop(serverComponents);
+                contextHandler.removeBean(serverComponents);
+                contextHandler.removeEventListener(this);
             }
         });
 
+        servletContext.setAttribute(WEBSOCKET_COMPONENTS_ATTRIBUTE, serverComponents);
```

Comment:
Do you need to remove this attribute when stopping?

Reply:
Its currently being removed automatically, but it wouldn't hurt to also remove this when stopping.
I will add a test case for this as well.